File: routedeck_core/contracts/events.py
from __future__ import annotations

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, model_validator

from .failures import RouteDeckFailure
from .projection import PublicEntityHandle, PublicValue


class _FrozenContract(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")


class RouteDeckEventType(StrEnum):
    SESSION_CREATED = "session_created"
    PROJECTION_CHANGED = "projection_changed"
    NAVIGATION_CHANGED = "navigation_changed"
    OPERATION_CHANGED = "operation_changed"
    PRIVATE_FORM_CHANGED = "private_form_changed"
    TURN_FINALIZED = "turn_finalized"
    TURN_INTERRUPTED = "turn_interrupted"


class PublicEventPayload(_FrozenContract):
    node_id: str | None = None
    operation_id: str | None = None
    request_id: str | None = None
    status_code: str | None = None
    entity_handles: tuple[PublicEntityHandle, ...] = ()
    details: tuple[PublicValue, ...] = ()
    failure: RouteDeckFailure | None = None


class CanonicalRouteDeckEvent(_FrozenContract):
    """Canonical Task 4 event; the root legacy event remains a compatibility API."""

    event_id: str = Field(min_length=1)
    cursor: int = Field(ge=1)
    event_type: RouteDeckEventType
    session_id: str = Field(min_length=1)
    session_version: int = Field(ge=0)
    projection_version: int | None = Field(default=None, ge=0)
    created_at: datetime
    payload: PublicEventPayload

    @model_validator(mode="after")
    def _aware_timestamp(self) -> CanonicalRouteDeckEvent:
        if self.created_at.tzinfo is None:
            raise ValueError("created_at must be timezone-aware")
        return self

    @property
    def event_cursor(self) -> int:
        return self.cursor

    @property
    def type(self) -> RouteDeckEventType:
        return self.event_type


RouteDeckEvent = CanonicalRouteDeckEvent


class EventPage(_FrozenContract):
    events: tuple[CanonicalRouteDeckEvent, ...]
    next_cursor: int = Field(ge=0)
    has_more: bool
    reset_required: bool = False
    retained_from_cursor: int | None = Field(default=None, ge=0)

    @model_validator(mode="after")
    def _reset_contract(self) -> EventPage:
        if self.reset_required:
            if self.events or self.has_more or self.retained_from_cursor is None:
                raise ValueError(
                    "reset-required event pages contain only the retained cursor"
                )
        elif self.retained_from_cursor is not None:
            raise ValueError("retained_from_cursor is valid only for reset pages")
        return self


__all__ = [
    "CanonicalRouteDeckEvent",
    "EventPage",
    "PublicEventPayload",
    "RouteDeckEvent",
    "RouteDeckEventType",
]

File: packages/core/src/contracts/events.ts
import type {
  PublicEventPayload,
  PublicRouteDeckEvent as GeneratedRouteDeckEvent,
  RouteDeckEventType,
} from "./generated";
import { generatedObjectDescriptors } from "./generatedRuntime";
import strictJsonDecoders from "./json";
import operationDecoders from "./operations";
import type {
  RouteDeckPublicEntityHandle,
  RouteDeckPublicValue,
} from "./projection";
import projectionDecoders from "./projection";

const {
  decodeArray,
  decodeNullableString,
  expectEnum,
  expectInteger,
  expectIsoDate,
  expectRecord,
  expectString,
} = strictJsonDecoders;
const { decodeFailure } = operationDecoders;
const { decodeEntity, decodePublicValue } = projectionDecoders;

export interface RouteDeckEventPayload
  extends Omit<PublicEventPayload, "details" | "entity_handles"> {
  details: RouteDeckPublicValue[];
  entity_handles: RouteDeckPublicEntityHandle[];
}

export interface RouteDeckEvent
  extends Omit<
    GeneratedRouteDeckEvent,
    "payload" | "projection_version" | "session_id"
  > {
  payload: RouteDeckEventPayload;
  projection_version: number | null;
}

const EVENT_TYPES = new Set<RouteDeckEventType>([
  "session_created",
  "projection_changed",
  "navigation_changed",
  "operation_changed",
  "private_form_changed",
  "turn_started",
  "turn_finalized",
  "turn_interrupted",
]);

export function decodeEvent(value: unknown): RouteDeckEvent {
  const record = expectRecord(
    value,
    "$event",
    generatedObjectDescriptors.PublicRouteDeckEvent,
  );
  const eventType = expectEnum(
    record.event_type,
    "$event.event_type",
    EVENT_TYPES,
  );
  const payloadRecord = expectRecord(
    record.payload,
    "$event.payload",
    generatedObjectDescriptors.PublicEventPayload,
  );
  return {
    created_at: expectIsoDate(record.created_at, "$event.created_at"),
    cursor: expectInteger(record.cursor, "$event.cursor", 1),
    event_id: expectString(record.event_id, "$event.event_id"),
    event_type: eventType,
    payload: {
      node_id: decodeNullableString(
        payloadRecord.node_id === undefined ? null : payloadRecord.node_id,
        "$event.payload.node_id",
      ),
      operation_id: decodeNullableString(
        payloadRecord.operation_id === undefined ? null : payloadRecord.operation_id,
        "$event.payload.operation_id",
      ),
      request_id: decodeNullableString(
        payloadRecord.request_id === undefined ? null : payloadRecord.request_id,
        "$event.payload.request_id",
      ),
      status_code: decodeNullableString(
        payloadRecord.status_code === undefined ? null : payloadRecord.status_code,
        "$event.payload.status_code",
      ),
      entity_handles: decodeArray(
        payloadRecord.entity_handles === undefined ? [] : payloadRecord.entity_handles,
        "$event.payload.entity_handles",
        decodeEntity,
      ),
      details: decodeArray(
        payloadRecord.details === undefined ? [] : payloadRecord.details,
        "$event.payload.details",
        decodePublicValue,
      ),
      failure:
        payloadRecord.failure === undefined || payloadRecord.failure === null
          ? null
          : decodeFailure(payloadRecord.failure, "$event.payload.failure"),
    },
    projection_version:
      record.projection_version === undefined || record.projection_version === null
        ? null
        : expectInteger(record.projection_version, "$event.projection_version", 0),
    session_version: expectInteger(
      record.session_version,
      "$event.session_version",
      0,
    ),
  };
}
